gprofiler/heartbeat.py

```diff
@@ -0,0 +1,627 @@
+                return False
+                
+        except Exception as e:
+            logger.error(f"Failed to send command completion for {command_id}: {e}")
+            return False
+
+    
+    
+    def mark_command_executed(self, command_id: str):
+        """Mark a command as executed (in-memory)"""
+        self.executed_command_ids.add(command_id)
+        
+        # Cleanup old command IDs if we exceed the limit
+        if len(self.executed_command_ids) > self.max_command_history:
+            self._cleanup_old_command_ids()
+        
+        logger.debug(f"Marked command ID {command_id} as executed")
+    
+    def _cleanup_old_command_ids(self):
+        """Remove old command IDs to prevent memory growth"""
+        try:
+            # Keep only the most recent commands (this is a simple approach)
+            # In production, you might want to implement time-based cleanup
+            if len(self.executed_command_ids) > self.max_command_history:
+                # Convert to list, sort, and keep the last max_command_history items
+                command_list = list(self.executed_command_ids)
+                # Since UUIDs don't sort chronologically, we'll just remove some arbitrary ones
+                # In a real implementation, you'd want to track timestamps
+                commands_to_keep = command_list[-self.max_command_history:]
+                self.executed_command_ids = set(commands_to_keep)
+                logger.info(f"Cleaned up command ID history in memory, keeping {len(self.executed_command_ids)} entries")
+        except Exception as e:
+            logger.warning(f"Failed to cleanup old command IDs: {e}")
+
+
+class DynamicGProfilerManager:
+    """Manager for dynamically starting/stopping gProfiler instances based on server commands"""
+    
+    def __init__(self, base_args: configargparse.Namespace, heartbeat_client: HeartbeatClient):
+        self.base_args = base_args
+        self.heartbeat_client = heartbeat_client
+        self.current_gprofiler: Optional['GProfiler'] = None
+        self.current_thread: Optional[threading.Thread] = None
+        self.stop_event = threading.Event()
+        self.heartbeat_interval = 30  # seconds
+    
+    def start_heartbeat_loop(self):
+        """Start the main heartbeat loop"""
+        logger.info("Starting heartbeat loop...")
+        
+        while not self.stop_event.is_set():
+            try:
+                # Send heartbeat and check for commands
+                command_response = self.heartbeat_client.send_heartbeat()
+                
+                if command_response and command_response.get("profiling_command"):
+                    profiling_command = command_response["profiling_command"]
+                    command_id = command_response["command_id"]
+                    command_type = profiling_command.get("command_type", "start")
+                    
+                    logger.info(f"Received profiling command: {profiling_command}")
```

Comment:
This is redundant with the line 119. It would be better to turn one of them into debug